Update tests for NavigatorID

foolip opened the pull request:
Aligns with these spec changes:
https://www.w3.org/Bugs/Public/show_bug.cgi?id=27796
https://github.com/whatwg/html/pull/78
https://github.com/whatwg/html/pull/199
https://github.com/whatwg/html/pull/207
https://github.com/whatwg/html/pull/220
https://github.com/whatwg/html/pull/1017
https://github.com/whatwg/html/pull/1870

foolip commented:
How should the NavigatorID.worker.js variant be run? I think there'll be lots of failures, and I wanted to use that to maybe change the spec again.

foolip commented:
For the window tests, I have verified that they all pass, except for Edge failing on the navigator.vendor test, where I expect "Google Inc." If this PR is merged I'll file an Edge issue to see if they'd go along with this. If not, another compatibility mode would be needed.

Ms2ger commented:
You may want to check if workers/interfaces/WorkerUtils/navigator/window-only.worker.js is redundant now.

Ms2ger commented:
http://w3c-test.org/submissions/3881/html/webappapis/system-state-and-capabilities/the-navigator-object/NavigatorID.worker
http://w3c-test.org/submissions/3881/workers/interfaces/WorkerUtils/navigator/window-only.worker

foolip commented:
Oh, so [wptserve does it](https://github.com/w3c/wpt-tools/blob/5b4013ffae79602a0a633b6551d247bccb49ac8a/serve/serve.py#L36), filed https://crbug.com/653514

window-only.worker.js is indeed now redundant, will remove.

foolip commented:
In http://w3c-test.org/submissions/3881/html/webappapis/system-state-and-capabilities/the-navigator-object/NavigatorID.worker, these tests fail:
 * Chrome 55: `appCodeName` and `product`
 * Edge 14: `product`, `productSub`, `vendor` and `vendorSub`
 * Firefox 51:  `appCodeName` and `product`
 * Safari 10:  `appCodeName`, `product`, `productSub`, `vendor` and `vendorSub`

Inverted:
* `appCodeName`: all except Edge fail
* `product`: all fail
* `productSub`, `vendor` and `vendorSub`: Edge and Safari fail

foolip commented:
I've submitted a spec PR to expose `appCodeName` and `product` and will now update the tests.

foolip committed:
Update tests for NavigatorID

Aligns with these spec changes:
https://www.w3.org/Bugs/Public/show_bug.cgi?id=27796
https://github.com/whatwg/html/pull/78
https://github.com/whatwg/html/pull/199
https://github.com/whatwg/html/pull/207
https://github.com/whatwg/html/pull/220
https://github.com/whatwg/html/pull/1017
https://github.com/whatwg/html/pull/1870

foolip commented:
This will leave Edge and WebKit with failures, will file bugs when spec and tests are merged.

foolip commented:
@domenic, can you review?

domenic commented on html/webappapis/system-state-and-capabilities/the-navigator-object/NavigatorID.js:
```diff
@@ -60,4 +53,51 @@ function run_test() {
+  // https://www.w3.org/Bugs/Public/show_bug.cgi?id=27820
+
+  test(function() {
+    if ("window" in self && navigator.userAgent.indexOf("WebKit") == -1) {
```
This file uses several different ways of testing navigator compatibility mode, not all of which are compatible. E.g. here Gecko means "does not have the string WebKit", where above Gecko means "does not have the string WebKit or the string Chrome".

It would be good to factor this out into a variable (in the test file) that is one of "Chrome", "WebKit", or "Gecko", and then the test code proper could say `if ("window" in self && navigatorCompatibilityMode === "Gecko")`. Then you wouldn't need the spec quotes as much.

domenic commented on html/webappapis/system-state-and-capabilities/the-navigator-object/NavigatorID.js:
```diff
@@ -23,31 +22,25 @@ function run_test() {
-      // Otherwise it should exist and return false.
-      else {
-        assert_false(navigator.taintEnabled());
+      if (navigator.userAgent.indexOf("Chrome") != -1 ||
```
Nit: this could be `navigator.userAgent.includes("Chrome")` for clarity.

foolip committed:
fixup! Update tests for NavigatorID
